pkgs/development/python-modules/apache-airflow/update-providers.py

```diff
@@ -0,0 +1,226 @@
+#! /usr/bin/env python3
+
+from itertools import chain
+import json
+import logging
+from pathlib import Path
+import os
+import re
+import subprocess
+import sys
+from typing import Dict, List, Optional, Set, TextIO
+from urllib.request import urlopen
+from urllib.error import HTTPError
+import yaml
+
+PKG_SET = "pkgs.python3Packages"
+
+# If some requirements are matched by multiple or no Python packages, the
+# following can be used to choose the correct one
+PKG_PREFERENCES = {
+    "dnspython": "dnspython",
+    "google-api-python-client": "google-api-python-client",
+    "psycopg2-binary": "psycopg2",
+    "requests_toolbelt": "requests-toolbelt",
+}
+
+# Requirements missing from the airflow provider metadata
+EXTRA_REQS = {
+    "sftp": ["pysftp"],
+}
+
+
+def get_version():
+    with open(os.path.dirname(sys.argv[0]) + "/default.nix") as fh:
+        # A version consists of digits, dots, and possibly a "b" (for beta)
+        m = re.search('version = "([\\d\\.b]+)";', fh.read())
+        return m.group(1)
+
+
+def get_file_from_github(version: str, path: str):
+    with urlopen(
+        f"https://raw.githubusercontent.com/apache/airflow/{version}/{path}"
+    ) as response:
+        return yaml.safe_load(response)
+
+
+def repository_root() -> Path:
+    return Path(os.path.dirname(sys.argv[0])) / "../../../.."
+
+
+def dump_packages() -> Dict[str, Dict[str, str]]:
+    # Store a JSON dump of Nixpkgs' python3Packages
+    output = subprocess.check_output(
+        [
+            "nix-env",
+            "-f",
+            repository_root(),
+            "-qa",
+            "-A",
+            PKG_SET,
+            "--arg",
+            "config",
+            "{ allowAliases = false; }",
+            "--json",
+        ]
+    )
+    return json.loads(output)
+
+
+def remove_version_constraint(req: str) -> str:
+    return re.sub(r"[=><~].*$", "", req)
+
+
+def name_to_attr_path(req: str, packages: Dict[str, Dict[str, str]]) -> Optional[str]:
+    if req in PKG_PREFERENCES:
+        return f"{PKG_SET}.{PKG_PREFERENCES[req]}"
+    attr_paths = []
+    names = [req]
+    # E.g. python-mpd2 is actually called python3.6-mpd2
+    # instead of python-3.6-python-mpd2 inside Nixpkgs
+    if req.startswith("python-") or req.startswith("python_"):
+        names.append(req[len("python-") :])
```

Comment:
Matching against name of python packages is brittle. Can we just rely on attr path?

Reply:
Also copied from home-assistant and we arrived iteratively at this solution. It is not brittle.

Reply:
Yep, I tried to base this as much as possible on the script @mweinelt refers to (https://github.com/NixOS/nixpkgs/blob/master/pkgs/servers/home-assistant/parse-requirements.py)